### PR TITLE
get vector variables out of a model properly

### PR DESCRIPTION
**Pull request by bqpd**
related to #822

### CODE REVIEW

**Comment by bqpd** [comment]
@whoburg, ready for review

File: gpkit/constraints/set.py [hunk 1]
```diff
@@ -60,14 +62,28 @@ def __getitem__(self, key):
             return list.__getitem__(self, key)
         else:
             from ..nomials import Variable
-            variables = [Variable(**key.descr) for key in self.varkeys[key]]
+            variables = [Variable(**k.descr) for k in self.varkeys[key]]
```
Comment (whoburg) [inline gpkit/constraints/set.py]: good catch

File: gpkit/constraints/set.py [hunk 2]
```diff
@@ -60,14 +62,28 @@ def __getitem__(self, key):
...
             if len(variables) == 1:
                 return variables[0]
-            else:
-                variables.sort(key=_sort_by_num_models)
-                variable = variables[0]
-                # note: doesn't work for vector variables
-                return variable
+            elif variables[0].key.idx and variables[0].key.shape:
+                # maybe it's all one vector variable!
+                vector = np.full(variables[0].key.shape, np.nan, dtype="object")
```
Comment (whoburg) [inline gpkit/constraints/set.py]: is `variables[0].key` guaranteed to have a `shape` attribute?

File: gpkit/constraints/set.py [hunk 3]
```diff
@@ -60,14 +62,28 @@ def __getitem__(self, key):
...
+                for variable in variables:
+                    testdict = dict(variable.key.descr)
+                    del testdict["idx"]
+                    if testdict == goaldict:
+                        vector[variable.key.idx] = variable
+                    else:
+                        raise ValueError("multiple variables are called '%s';"
+                                         " use variable_byname('%s') to see"
+                                         " all of them."
+                                         % (key, key))
```
Comment (whoburg) [inline gpkit/constraints/set.py]: cleaner to just `break` here, which will jump right into the identical raise statement outside the loop?

**Comment by whoburg** [comment]
review complete. And I think this needs a unit test before merging.

File: gpkit/constraints/set.py (hunk 2 above)
Comment (bqpd) [inline gpkit/constraints/set.py]: Yes, and it'll be none if it's not an ArrayVariable

File: gpkit/constraints/set.py (hunk 3 above)
Comment (bqpd) [inline gpkit/constraints/set.py]: derp, indeed.

**Comment by bqpd** [comment]
@whoburg, responded to code review and added test

**Comment by bqpd** [comment]
retest this please

File: gpkit/constraints/set.py (hunk 2 above)
Comment (whoburg) [inline gpkit/constraints/set.py]: what about after we address #844?

**Comment by whoburg** [comment]
retest this please

File: gpkit/tests/t_constraints.py [hunk 4]
```diff
@@ -12,6 +12,22 @@
...
+        with self.assertRaises(ValueError):
+            _ = Model(x, [x >= 1, x_ >= 1])["x"]
+        with self.assertRaises(ValueError):
+            _ = Model(x, [x >= 1, xv >= 1])["x"]
+        self.assertTrue(all(Model(xv.prod(), [xv >= 1])["x"] == xv))
+        with self.assertRaises(ValueError):
+            _ = Model(xv.prod(), [xv >= 1, xv_ >= 1])["x"]
+        with self.assertRaises(ValueError):
+            _ = Model(xv.prod(), [xv >= 1, x_ >= 1])["x"]
+
```
Comment (whoburg) [inline gpkit/tests/t_constraints.py]: great test

**Comment by whoburg** [comment]
this looks like a good improvement, let's merge it and deal with any issues related to #844 when fixing that issue.

**Pull request merged**